lib/shoppertrak_api_client.py

```diff
@@ -0,0 +1,161 @@
+    """Class for querying the ShopperTrak API for location visits data"""
+
+    def __init__(self, username, password):
+        self.logger = create_log("shoppertrak_api_client")
+        self.base_url = os.environ["SHOPPERTRAK_API_BASE_URL"]
+        self.auth = HTTPBasicAuth(username, password)
+        self.max_retries = int(os.environ["MAX_RETRIES"])
+        self.today_str = datetime.now(pytz.timezone("US/Eastern")).date().isoformat()
+
+    def query(self, endpoint, date_str, query_count=1):
+        full_url = self.base_url + endpoint
+
+        self.logger.info(f"Querying {endpoint} for {date_str} data")
+        try:
+            response = requests.get(
+                full_url,
+                params={"date": date_str, "increment": "15", "total_property": "N"},
+                auth=self.auth,
+            )
+            response.raise_for_status()
+        except RequestException as e:
+            self.logger.error(f"Failed to retrieve response from {full_url}: {e}")
+            raise ShopperTrakApiClientError(
+                f"Failed to retrieve response from {full_url}: {e}"
+            ) from None
+
+        response_root = self._check_response(response.text)
+        if response_root is not None:
+            return response_root
+        elif query_count < self.max_retries:
+            self.logger.info("Waiting 5 minutes and trying again")
+            time.sleep(300)
+            return self.query(endpoint, date_str, query_count + 1)
+        else:
+            self.logger.error(
+                f"Reached max retries: sent {self.max_retries} queries with no response"
+            )
+            raise ShopperTrakApiClientError(
+                f"Reached max retries: sent {self.max_retries} queries with no response"
+            )
+
+    def parse_response(self, xml_root, input_date_str, is_recovery_mode=False):
+        rows = []
+        for site_xml in xml_root.findall("site"):
+            site_str = self._get_xml_str(site_xml, "siteID")
+            for date_xml in site_xml.findall("date"):
+                date_str = self._get_xml_str(date_xml, "dateValue")
+                if date_str != input_date_str:
+                    self.logger.error(
+                        f"Request date does not match response date.\nRequest date: "
+                        f"{input_date_str}\nResponse date: {date_str}"
+                    )
+                    raise ShopperTrakApiClientError(
+                        f"Request date does not match response date.\nRequest date: "
+                        f"{input_date_str}\nResponse date: {date_str}"
+                    )
+                for entrance_xml in date_xml.findall("entrance"):
+                    entrance_val = self._get_xml_str(entrance_xml, "name")
+                    if entrance_val:
+                        entrance_val = self._cast_str_to_int(entrance_val.lstrip("EP"))
+                    for traffic_xml in entrance_xml.findall("traffic"):
```

Comment:
Personal preference thing here but it gets harder to follow the nesting as this gets deeper, might be easier to read refactored out into smaller methods?

Reply:
Yeah I thought about this but felt it was too confusing to include a list of methods each containing a for loop that needed to return both an xml root and a value. Instead, I split it off into two methods and added method descriptions which hopefully makes it easier to follow.